plugin/storage/cassandra/spanstore/reader.go

```diff
@@ -177,26 +240,119 @@ func (s *SpanReader) FindTraces(queryParams *spanstore.TraceQueryParameters) ([]
+	results := dbmodel.UniqueTraceIDs{}
+
+	minDurationMicros := (traceQuery.DurationMin.Nanoseconds() / int64(time.Microsecond/time.Nanosecond))
+	maxDurationMicros := (traceQuery.DurationMax.Nanoseconds() / int64(time.Microsecond/time.Nanosecond))
+
+	// See writer.go:indexByDuration  for how this is indexed
+	// This is indexed in hours since epoch, converted to seconds
+	// TODO encapsulate this calculation
+	startTimeSeconds := (traceQuery.StartTimeMin.UnixNano() / int64(time.Hour))
+	endTimeSeconds := (traceQuery.StartTimeMax.UnixNano() / int64(time.Hour))
+
+	for timeBucket := endTimeSeconds; timeBucket >= startTimeSeconds; timeBucket-- {
+		query := s.session.Query(
+			queryByDuration,
+			timeBucket,
+			traceQuery.ServiceName,
+			traceQuery.OperationName,
+			minDurationMicros,
+			maxDurationMicros,
+			traceQuery.NumTraces)
+		t, err := s.executeQuery(query, s.metrics.queryDurationIndex)
+		if err != nil {
+			return nil, err
+		}
+
+		for traceID := range t {
+			results[traceID] = struct{}{}
+			if len(results) == traceQuery.NumTraces {
+				break
+			}
+		}
+	}
+	return results, nil
+}
+
+func (s *SpanReader) queryByServiceNameAndOperation(tq *spanstore.TraceQueryParameters) (dbmodel.UniqueTraceIDs, error) {
+	query := s.session.Query(queryByServiceAndOperationName, tq.ServiceName, tq.OperationName, model.TimeAsEpochMicroseconds(tq.StartTimeMin), model.TimeAsEpochMicroseconds(tq.StartTimeMax), tq.NumTraces)
+	return s.executeQuery(query, s.metrics.queryServiceOperationIndex)
+}
+
+func (s *SpanReader) queryByService(tq *spanstore.TraceQueryParameters) (dbmodel.UniqueTraceIDs, error) {
+	query := s.session.Query(queryByServiceName, tq.ServiceName, model.TimeAsEpochMicroseconds(tq.StartTimeMin), model.TimeAsEpochMicroseconds(tq.StartTimeMax), tq.NumTraces)
+	return s.executeQuery(query, s.metrics.queryServiceNameIndex)
+}
+
+// executeQuery returns a list of type TraceID
+func (s *SpanReader) executeQuery(query cassandra.Query, tableMetrics *casMetrics.Table) (dbmodel.UniqueTraceIDs, error) {
 	start := time.Now()
-	q := s.session.Query(query.QueryString()).Bind(query.Parameters...)
-	i := q.Consistency(s.consistency).Iter()
-	var spans []dbmodel.Span
+	i := query.Consistency(s.consistency).Iter()
+	retMe := dbmodel.UniqueTraceIDs{}
 	var traceID dbmodel.TraceID
-	var spanID int64
-	for i.Scan(&traceID, &spanID) {
-		traceAndSpanID := dbmodel.Span{
-			TraceID: traceID,
-			SpanID:  spanID,
-		}
-		spans = append(spans, traceAndSpanID)
+	for i.Scan(&traceID) {
+		retMe[traceID] = struct{}{}
 	}
 	err := i.Close()
 	tableMetrics.Emit(err, time.Since(start))
 	if err != nil {
-		s.logger.Error("Failed to exec query", zap.String("query", query.QueryString()), zap.Any("params", query.Parameters), zap.Error(err))
+		s.logger.Error("Failed to exec query", zap.Error(err))
 		return nil, err
 	}
-	return dbmodel.GetUniqueTraceIDs(spans), nil
```

Comment:
is this GetUniqueTraceIDs function still used anywhere? If not, let's remove it.